grape.py

```diff
@@ -48,16 +48,32 @@ def grape(H0, Hk, u_0, rho_0, C, T, alpha, target="trace_real", epsilon=1e-6, ma
             break
 
         # last phi
-        phi = np.trace(np.dot( C.T.conjugate(), rhoj[-1] ))
+        phi = np.trace(C.T.conj() @ rhoj[-1])
+        if target == "trace_real":
+            pass
+        elif target == "trace_both":
+            phi = np.real(phi)
+        elif target == "abs":
+            phi = phi*phi
+        else:
+            raise ValueError("target function not supported")
 
         # calculate update_matrix and update u_kj, step to optimization
         update_matrix = None
         if target == "trace_real":
             update_matrix = gradient(lambdaj, rhoj, delta_t, Hk)
         elif target == "trace_both":
-            pass
+            lx = np.real(lambdaj)

```

Comment:
Base on the review before, consider how to calculate phi_1 (hint: think about how to decompose C)